### DISCUSSION TITLE
Add setting to allow folding function signature separately from body

### DISCUSSION THREAD

**Discussion by pedrorrivero** [microsoft/pylance-release#3913] [Enhancement]
Type: <b>Bug</b>

# Behaviour
## Expected vs. Actual

When the Python plugin is disabled, function/class bodies in `.py` files can be folded from the end of a multiline signature (i.e. at `:`); while the multiline signature can also be folded separetly at the same level. Enabling Python, breaks this nice behavior, only allwoing to fold the entire function from the first line of the signature (i.e. at `def`/`class`). 

This is, enabling Python joins signtaures and the bodies in a single foldable, instead of keeping them separate.

## Steps to reproduce:

1. Disable Python
2. Open `.py` file with a multiline function/class signature
3. Navigate with the cursor to the left edge of the file panel
4. Notice two arrows: one to fold the signature at `def`/`class` height, and another to fold the body at `:` height.
5. Enable Python
6. Close the `.py` file
6. Restart VS Code
8. Reopen the `.py` file
9. Notice that now there is only one arrow on the left edge of the file panel at `def`/`class` height which folds both the signature and body

# Diagnostic data

-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.9
-   Type of virtual environment used (e.g. conda, venv, virtualenv, etc.): Venv
-   Value of the `python.languageServer` setting: Default

<details>

<summary>Output for <code>Python</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Python</code>)
</summary>

<p>

```
XXX
```

</p>
</details>

<details>

<summary>User Settings</summary>

<p>

```

languageServer: "Pylance"

```

</p>
</details>

Extension version: 2022.14.0
VS Code version: Code 1.71.2 (Universal) (74b1f979648cc44d385a2286793c226e611f59e7, 2022-09-14T21:05:37.721Z)
OS version: Darwin x64 21.6.0
Modes:
Sandboxed: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz (16 x 2400)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|2, 2, 2|
|Memory (System)|32.00GB (0.97GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>

**Comment by pedrorrivero** [comment]
## Python disabled

### Expanded
<img width="779" alt="Screen Shot 2022-09-23 at 11 22 22 AM" src="https://user-images.githubusercontent.com/52980156/192023355-7a9fdc9d-a3c7-47e5-b37a-54cb8d792546.png">

### Collapse signature
<img width="776" alt="Screen Shot 2022-09-23 at 11 22 55 AM" src="https://user-images.githubusercontent.com/52980156/192023426-875341df-5be5-49a7-98db-37e0ee71ea99.png">

### Collapse body
<img width="779" alt="Screen Shot 2022-09-23 at 11 22 40 AM" src="https://user-images.githubusercontent.com/52980156/192023493-2954438b-a43b-4cd3-9315-657bb2920228.png">

### Collapse all
<img width="264" alt="Screen Shot 2022-09-23 at 11 23 06 AM" src="https://user-images.githubusercontent.com/52980156/192023562-677a06cd-edf7-4420-accf-1348dcf099c8.png">


## Python enabled

### Expanded
<img width="780" alt="Screen Shot 2022-09-23 at 11 21 38 AM" src="https://user-images.githubusercontent.com/52980156/192023167-89b8e3b2-186b-4fd7-875f-08de2d685a17.png">

### Collapse all
<img width="233" alt="Screen Shot 2022-09-23 at 11 21 53 AM" src="https://user-images.githubusercontent.com/52980156/192023299-f665f97e-c307-4a07-96e3-bbb45dadb54f.png">

**Comment by debonte** [comment]
We previously discussed this in https://github.com/microsoft/pylance-release/issues/2269. Take a look at the three options I described in [this comment](https://github.com/microsoft/pylance-release/issues/2269#issuecomment-1018064327). 

Option 1 is your preferred behavior.
Option 2 is the current behavior.
Option 3 is an approach that I think may make everyone happy.

Let me know what you think.

**Comment by pedrorrivero** [comment]
Thanks for the quick reply @debonte ! I agree with you, option 3 is a good intermediate approach. 

The important thing imho is to be able to keep the signature visible. Only concern with option 3 really is that it would mess up the levels so that shortcuts will not be consistent (e.g. folding level 2 would mean folding the body in some functions, and folding something internal in other cases depending on whether the signature is multiline or not)

That is why I would personally advocate for option 1. Plus the overall underlying data structure seems more natural to me, so easier to understand as a user.

Thanks for putting all this together 😉

**Comment by debonte** [comment]
> Only concern with option 3 really is that it would mess up the levels so that shortcuts will not be consistent (e.g. folding level 2 would mean folding the body in some functions, and folding something internal in other cases depending on whether the signature is multiline or not)

I'm not sure I understand.

To recap, option 3 means that in the following screenshot, folding line 22 collapses the body (leaving the parameter list visible) and folding line 18 collapses the entire function definition (arguments and body):
![image](https://user-images.githubusercontent.com/12820898/197071583-6e18f7a1-5c97-4155-8157-d0a57dd141b7.png)

I think your concern is that with single-line function headers, users might think that the next `>` icon below the `def` will collapse the function body? In the example below, folding line 189 collapses the entire function. If we go with option 3, are you concerned that users will think that folding line 190 would collapse the body of the function, simply because it's the next folding icon below the `def` line?
![image](https://user-images.githubusercontent.com/12820898/197071919-44ca61c1-f05a-4f41-91e2-9a943dc7b823.png)

If that's your concern, I'm not sure how to address it. That would have been a problem with the old (2021, aka option 1) behavior as well. Also, this would seem like an argument for sticking with the current behavior -- a single, consistent folding location for functions.

If I'm misunderstanding your point, could you show me some code snippets to explain what you're describing?

**Comment by debonte** [comment]
Some other thoughts:
1. What if we changed the text of the folding line when folding functions with multiline signatures? So if you folded line 18 of `my_func` above, and it folded the entire function, the text shown on line 18 would be `def my_func(a, b, c)` instead of `def my_func(`. This would require vscode to support LSP's `FoldingRange.collapsedText` field. See https://github.com/microsoft/vscode/issues/70794.
2. We could have a config setting to switch between 2+ function folding behaviors.

**Comment by pedrorrivero** [comment]
Thanks so much for the detailed response!

This is what I mean. Take the following code snippet as an example, folding points are labeled alphabetically:
```python
def single_line():          # A
    if 0 == 1:              # B
        pass

def multi_line(             # C
    *args,
):                          # D
    pass
```

Folding level 1 means:
- Under option 1 → `#A`, `#C` and `#D` (exposed)
- Under option 3 → `#A` and `#C` (`#D` is hidden)

Folding level 2 means:
- Under option 1 →  `#B`
- Under option 3 → `#B` and `#D`

Notice how folding level 1 is roughly equivalent and equivalent; with the added choice of displaying the signature under option 1 afterwards.

However, folding level 2 under option 3 messes the structure of the document, since one function will get its body folded and the other only an internal `if` statement. And the situation gets worse if you start considering these as methods within a class (which, if the class is large, I like to fold automatically with a key binding for quickly looking through it).

You can toy around with this idea and notice that the real issue with option 3 is that it assigns different precedence levels to things which are structurally at the same "height": throwing the body under the signature instead of keeping them aligned in the hierarchy.

---

> We could have a config setting to switch between 2+ function folding behaviors

This would be a nice way to end the controversy, let users choose what they want.

> What if we changed the text of the folding line when folding functions with multiline signatures?

In which case this would help remind them what their selected choice is 🙂  but it is not a huge concern of mine. I think people can get a grasp of the behavior by simply clicking a few times

**Comment by debonte** [comment]
Thanks. I'm embarrassed to say that I wasn't familiar with the `Fold Level #` commands until now.

I'll need to think about this some more and play with the different behaviors. Thanks for educating me. :)

**Comment by debonte** [comment]
For future reference, here are some screenshots I captured using the three options on the following code:

```python
def single_line():          # A
    if 0 == 1:              # B
        pass

def multi_line(             # C
    *args,
):                          # D
    if 0 == 1:
        pass

class Class:
    def single_line(self):  # A
        if 0 == 1:          # B
            pass

    def multi_line(         # C
        self,           
        *args,
    ):                      # D
        pass

def another():
    pass
```

Folding level 1 (option 1 on left, option 2 center, option 3 on right):
![FoldingLevel1](https://user-images.githubusercontent.com/12820898/213793642-2808f1b4-b224-4947-9295-7e103b3c75e3.png)

Folding level 2:
![FoldingLevel2](https://user-images.githubusercontent.com/12820898/213793663-9950d9b9-5f76-4671-a01e-b1943b806468.png)

**Comment by debonte** [comment]
I agree that option 3's behavior with the `Fold Level #` commands is weird and I think it rules that out as an option.

If there's sufficient interest, we can add a setting to choose between options 1 and 2 for folding functions (and classes) with multi-line signatures/headers. However, given that no one else has upvoted or commented on this issue, it's not clear that it's broadly needed.

@pedrorrivero, can you try this setting to see if it is acceptable to you as a workaround in the meantime?

```json
"[python]": {
    "editor.foldingStrategy": "indentation",
},
```

**Comment by debonte** [comment]
@pedrorrivero, ping? See my comment above.

**Comment by pedrorrivero** [comment]
Apologies, I've had very busy weeks. Let me take a look at this over the weekend! Thanks for pinging @debonte 🙂

**Comment by pedrorrivero** [comment]
This looks very nice indeed, for what I have been able to see it does exactly what we discussed for option 1, thanks a lot for taking a look and providing a solution @debonte.

Looking forward to that extra option in the future if there is interest 😄 Let me know if I can help in any other way

**Comment by pedrorrivero** [comment]
Quick update @debonte, the workaround you proposed fails for docstrings (i.e. they cannot be collapsed).

**Reply by debonte** [reply]
It's the default folding strategy provided by VS Code. We don't have any control over how it works. It's purely indentation based -- a line will be foldable if the line/lines below it are more indented.

**Reply by pedrorrivero** [reply]
Yeah, I figured... just wanted to let you know (and leave a record) that the solution does not work in general. In case someone else gets to this thread in the future.

**Comment by munael** [comment]
I think fine control on folding is not _exactly_ what we're looking for when raising this issue.

At least, in my case it was the need to see function signatures while functions are collapsed.

So as an extension to the current behavior, how about adding a configuration for:
1. Current behavior: Folding indicator starts at `def` and folds the entire definition. Only the first line is visible in the collapsed text.
2. Optional behavior: Folding indicator starts at `def` OR the end of the signature and folds only the _function body_. There's one single folding arrow. If this option is activated, function signatures are _never_ folded.

This configuration should be toggle-able in an open file without changing the global, workspace or folder settings json.

cc @debonte
